uitools/cpp/Esri/ArcGISRuntime/Toolkit/FloorFilterController.cpp

```diff
@@ -0,0 +1,698 @@
+      {
+        setSelectedLevelId({});
+        setSelectedFacilityId({});
+        setSelectedSiteId({});
+      }
+      // Assumption: if too zoomed out to see sites, also too zoomed out to see facilities
+      return;
+    }
+
+    if (!floorManager)
+      return;
+
+    // If the centerpoint is within a site's geometry, select that site.
+    // This code gracefully skips selection if there are no sites or no matching sites
+    auto sites = floorManager ? floorManager->sites() : QList<FloorSite*>{};
+    auto siteResult = std::find_if(std::cbegin(sites), std::cend(sites),
+                                   [&observedViewpoint](FloorSite* site)
+                                   {
+                                     if (site == nullptr)
+                                       return false;
+
+                                     auto extent = site->geometry().extent();
+                                     if (extent.isEmpty())
+                                       return false;
+
+                                     return GeometryEngine::intersects(extent, observedViewpoint.targetGeometry());
+                                   });
+
+    if (siteResult != std::cend(sites))
+    {
+      setSelectedSiteId((*siteResult)->siteId());
+    }
+    else if (m_automaticSelectionMode == AutomaticSelectionMode::Always)
+    {
+      setSelectedSiteId({});
+    }
+
+    // Move on to facility selection. Default to map-authored Facility MinScale. If MinScale not specified or is 0, default to 1500.
+    targetScale = 0.0;
+    if (floorManager)
+    {
+      if (auto siteLayer = floorManager->facilityLayer())
+      {
+        targetScale = siteLayer->minScale();
+      }
+    }
+
+    if (targetScale == 0)
+    {
+      targetScale = 1500;
+    }
+
+    // If out of scale, stop here
+    if (observedViewpoint.targetScale() > targetScale)
+    {
+      return;
+    }
+
+    auto facilities = floorManager ? floorManager->facilities() : QList<FloorFacility*>{};
+    auto facilityResult = std::find_if(std::cbegin(facilities), std::cend(facilities),
+                                       [&observedViewpoint](FloorFacility* facility)
+                                       {
+                                         if (facility == nullptr)
+                                           return false;
+
+                                         auto extent = facility->geometry().extent();
+                                         if (extent.isEmpty())
+                                           return false;
+
+                                         return GeometryEngine::intersects(extent, observedViewpoint.targetGeometry());
+                                       });
+
+    if (facilityResult != std::cend(facilities))
+    {
+      setSelectedFacilityId((*facilityResult)->facilityId());
+    }
+    else if (m_automaticSelectionMode == AutomaticSelectionMode::Always)
+    {
+      setSelectedFacilityId({});
+    }
```

Comment:
By testing out mine and yours viewpoint implementation I found a difference. In yours, when the facility is not in view, the levels visibility are not getting cleared. 
If you agree, then you are missing the `setSelectedLevelId({})`

https://user-images.githubusercontent.com/91129623/156550392-de1f2a0e-631c-4a39-bb29-8cfbe91ddea4.mp4